src/persist/src/nemesis/progress.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+
+use timely::dataflow::ProbeHandle;
+use timely::progress::Antichain;
+use timely::PartialOrder;
+
+use crate::error::Error;
+use crate::future::{Future, FutureHandle};
+
+/// A mechanism for externally waiting on dataflow progress.
+#[derive(Clone, Debug)]
+pub struct DataflowProgress {
+    // When the tx side closes, it sets this bool to true before draining and
+    // closing the channel. This is to handle the possibility of a race where a
+    // new waiter is added after the channel drain but before its closed. For
+    // us, this means if the bool is true, we immediately fill the returned
+    // future with an error instead of attempting to write to the channel.
+    closed: Arc<AtomicBool>,
+    tx: crossbeam_channel::Sender<(u64, FutureHandle<()>)>,
+}
+
+impl DataflowProgress {
+    /// Returns a client for waiting on dataflow progress and the paired handle
+    /// for transmitting it.
+    pub fn new() -> (DataflowProgressHandle, DataflowProgress) {
+        let (tx, rx) = crossbeam_channel::unbounded();
+        let closed = Arc::new(AtomicBool::new(false));
+        (
+            DataflowProgressHandle {
+                core: Arc::new(Mutex::new(DataflowProgressCore {
+                    rx,
+                    closed: closed.clone(),
+                    waiters: BTreeMap::new(),
+                    frontier: Antichain::from_elem(0),
+                })),
+            },
+            DataflowProgress { tx, closed },
+        )
+    }
+
+    /// Returns a Future that is filled once the given timestamp is no longer in
+    /// advance of the paired dataflow probe's frontier (i.e. returns when
+    /// `!dataflow.less_equal(ts)`).
```

Comment:
this comment mentions `!less_equals` but lines 133 and 153 use `!less_than` -- im not sure which is right, just highlighting that they are different?

Reply:
Good catch, pretty sure the comment is right and the code is wrong, but I'll think about this again for a sec to be sure

Reply:
nope, the code is right and the comment is wrong. we do want to unblock waiting on a seal at timestamp T if some element of the fronter is == T. I literally spent 5 minutes carefully writing this stupid comment and still got it wrong...